odbc/src/api/functions.rs

```diff
@@ -1279,73 +1274,211 @@ pub unsafe extern "C" fn SQLFetch(statement_handle: HStmt) -> SqlReturn {
+
+    // if rows_fetched_ptr is null, it was not set by the user and can be ignored.
+    let has_rows_fetched_buffer = !stmt.attributes.read().unwrap().rows_fetched_ptr.is_null();
+
+    if has_rows_fetched_buffer {
+        *stmt.attributes.write().unwrap().rows_fetched_ptr = 0;
+    }
+
+    // keeps track of how many rows in the result set cause SqlReturn::ERROR when being handled.
+    let mut row_error_count = 0;
+
+    // Use `index` to figure out which buffer in the array of buffers to use.
+    for index in 0..rowset_size {
+        let move_to_next_result = {
+            let connection = must_be_valid!((*stmt.connection).as_connection());
+            match stmt.mongo_statement.write().unwrap().as_mut() {
+                Some(mongo_stmt) => mongo_stmt
+                    .next(connection.mongo_connection.read().unwrap().as_ref())
+                    .map_err(|e| e.into()),
+                None => Err(ODBCError::InvalidCursorState),
+            }
+        };
+
+        // if row_status_ptr is null, it was not set by the user and can be ignored.
+        // Otherwise, we need to set the row_status_buffer to the address that corresponds with the current row.
+        let row_status_buffer: *mut USmallInt =
+            if !stmt.attributes.read().unwrap().row_status_ptr.is_null() {
+                (stmt.attributes.read().unwrap().row_status_ptr as ULen
+                    + (index * size_of::<u16>())) as *mut USmallInt
+            } else {
+                null_mut()
+            };
 
-                match sql_return {
-                    SqlReturn::ERROR => return SqlReturn::ERROR,
-                    SqlReturn::SUCCESS_WITH_INFO => {
-                        success_with_info_encountered = true;
+        if let Ok((has_next, mut warnings_opt)) = move_to_next_result {
+            warnings_opt.iter().for_each(|warning| {
+                add_diag_with_function!(
+                    MongoHandleRef::from(statement_handle),
+                    ODBCError::GeneralWarning(warning.to_string()),
+                    function_name.to_string()
+                );
+            });
+
+            if !has_next {
+                // There are no more rows, so set the rest of Row Status Array buffers to SQL_ROW_NOROW if the rowset end has not been reached.
+                if !row_status_buffer.is_null() {
+                    for row_status_index in index..rowset_size {
+                        let row_status_buffer = (stmt.attributes.read().unwrap().row_status_ptr
+                            as ULen
+                            + (row_status_index * size_of::<u16>()))
+                            as *mut USmallInt;
+                        *row_status_buffer = definitions::SQL_ROW_NOROW;
                     }
-                    _ => {}
+                }
+
+                // break here to prevent NO_DATA from being returned if the rowset had at least one row in it.
+                if has_fetched_at_least_one_row {
```

Comment:
I need to break no matter what since there are no rows left. So it would have to be like this if we change it:
```
if !has_fetched_at_least_one_row {
return SqlReturn::NO_DATA;
}else{
  break;
}
```